api/v1/verticadb_webhook.go

```diff
@@ -1701,6 +1728,202 @@ func (v *VerticaDB) checkSandboxPrimary(allErrs field.ErrorList, oldObj *Vertica
+			err := field.Invalid(p,
+				newSBox.Shutdown,
+				fmt.Sprintf("shutdown must be false when adding sandbox %q",
+					newSBox.Name))
+			allErrs = append(allErrs, err)
+		}
+	}
+	for i := range v.Spec.Subclusters {
+		newSCluster := &v.Spec.Subclusters[i]
+		if _, foundInStatus := statusSClusterMap[newSCluster.Name]; !foundInStatus && newSCluster.Shutdown {
+			p := field.NewPath("spec").Child("subclusters").Index(i).Child("shutdown")
+			err := field.Invalid(p,
+				newSCluster.Shutdown,
+				fmt.Sprintf("shutdown must be false when adding subcluster %q",
+					newSCluster.Name))
+			allErrs = append(allErrs, err)
+		}
+	}
+	return allErrs
+}
+
+// findSclustersToUnsandboxreturn a map whose key is the pointer of a subcluster that is to be unsandboxed.
+// The value is the pointer to a sandbox where the subcluster lives in
+func (v *VerticaDB) findSclustersToUnsandbox(old runtime.Object) map[*Subcluster]*Sandbox {
+	oldObj := old.(*VerticaDB)
+	oldSubclusterInSandbox := oldObj.GenSubclusterSandboxMap()
+	newSuclusterInSandbox := v.GenSubclusterSandboxMap()
+	oldSubclusterMap := oldObj.GenSubclusterMap()
+	oldSandboxMap := oldObj.GenSandboxMap()
+	newSubclusterMap := v.GenSubclusterMap()
+	sclusterSboxMap := make(map[*Subcluster]*Sandbox)
+	for oldSubclusterName, oldSandboxName := range oldSubclusterInSandbox {
+		_, oldSubclusterInNewSandboxes := newSuclusterInSandbox[oldSubclusterName]
+		_, oldSubclusterInPersist := newSubclusterMap[oldSubclusterName]
+		// for unsandboxing, check shutdown field of the subcluster and sandbox
+		if !oldSubclusterInNewSandboxes && oldSubclusterInPersist {
+			oldSubcluster := oldSubclusterMap[oldSubclusterName]
+			oldSandbox := oldSandboxMap[oldSandboxName]
+			sclusterSboxMap[oldSubcluster] = oldSandbox
+		}
+	}
+	return sclusterSboxMap
+}
+
+// checkUnsandboxShutdownConditions ensures we will not unsandbox a subcluster that has shutdown set to true, or a subcluster
+// in a sandbox where the shutdown field for the sandbox is set to true
+func (v *VerticaDB) checkUnsandboxShutdownConditions(old runtime.Object, allErrs field.ErrorList) field.ErrorList {
+	statusSClusterIndexMap := v.GenStatusSClusterIndexMap()
+	subclustersToUnsandbox := v.findSclustersToUnsandbox(old)
+	subclusterIndexMap := v.GenSubclusterIndexMap()
+	sandboxIndexMap := v.GenSandboxIndexMap()
+	for oldSubcluster, oldSandbox := range subclustersToUnsandbox {
+		oldSubclusterIndex, found := statusSClusterIndexMap[oldSubcluster.Name]
+		if oldSubcluster.Shutdown || found && v.Status.Subclusters[oldSubclusterIndex].Shutdown {
+			i := subclusterIndexMap[oldSubcluster.Name]
+			p := field.NewPath("spec").Child("subclusters").Index(i)
+			err := field.Invalid(p,
+				oldSubcluster.Shutdown,
+				fmt.Sprintf("cannot unsandbox subcluster %q that has Shutdown field set to true",
+					oldSubcluster.Name))
+			allErrs = append(allErrs, err)
+			continue
+		}
+		if oldSandbox.Shutdown {
```

Comment:
For one sandbox, we probably reach here multiple times so we probably return duplicate errors.